kolena/_experimental/dataset/_dataset.py

```diff
@@ -273,23 +279,70 @@ def _iter_dataset_raw(
     )
 
 
-def iter_dataset(
+def _iter_dataset(
     name: str,
+    commit: str = None,
     batch_size: int = BatchSize.LOAD_SAMPLES.value,
 ) -> Iterator[pd.DataFrame]:
     """
     Get an iterator over datapoints in the dataset.
     """
-    for df_batch in _iter_dataset_raw(name, batch_size):
+    for df_batch in _iter_dataset_raw(name, commit, batch_size):
         yield _to_deserialized_dataframe(df_batch, column=COL_DATAPOINT)
 
 
 def fetch_dataset(
     name: str,
+    commit: str = None,
     batch_size: int = BatchSize.LOAD_SAMPLES.value,
 ) -> pd.DataFrame:
     """
     Fetch an entire dataset given its name.
     """
-    df_batches = list(iter_dataset(name, batch_size))
+    df_batches = list(_iter_dataset(name, commit, batch_size))
     return pd.concat(df_batches, ignore_index=True) if df_batches else pd.DataFrame()
+
+
+def _list_commits(name: str, desc: bool = False, offset: int = 0, limit: int = 50) -> ListCommitHistoryResponse:
+    """
+    Invoke the list-commits api.
+    """
+    request = ListCommitHistoryRequest(name=name, desc=desc, offset=offset, limit=limit)
+    response = krequests.put(Path.LIST_COMMITS, json=asdict(request))
+    response.raise_for_status()
+    return from_dict(ListCommitHistoryResponse, response.json())
+
+
+def _iter_commits(
+    name: str,
+    desc: bool = False,
+    limit: int = None,
+    page_size: int = 50,
+) -> Iterator[Tuple[int, List[CommitData]]]:
+    """
+    Get an iterator over the commit history of the dataset.
+    """
+    initial_response = _list_commits(name, desc=desc, limit=page_size)
+    total_commit_count = initial_response.total_count
+    if not limit:
+        limit = total_commit_count
+    yield total_commit_count, initial_response.records
+    current_count = len(initial_response.records)
+    while current_count < min(limit, total_commit_count):
+        yield total_commit_count, _list_commits(name, desc=desc, offset=current_count, limit=page_size).records
+        current_count += len(initial_response.records)
```

Comment:
why are we using `initial_response.records` here not the last request length? Also, should the while loop also use the `total` from latest response?

Reply:
Good catch, should increment with the new response.

As for using the `total_count` from the initial response, that's because in case of edits that happen in parallel to the `get_dataset_history()` call, the `total_count` may increase between pages and this while loop might end up being an infinite one.

Reply:
With offset and limit given every request, and that server respond includes latest total_count, why would this be infinite loop?

Reply:
Thought through a bit more, I think you are right and we should not expect infinite loop given that the commit history can only increase over time.
- With `descending=False`: there is no problem, as the `offset` will serve as the cursor on chronological order
- With `descending=True`: there might be situations where the same commit is retrieved multiple times if new commits are added

Reply:
For 2nd case, supposedly we can dedup on client side, with caveat we'll for sure miss the new commits, but I think in a way acceptable.